src/jit/morph.cpp

```diff
@@ -7737,12 +7737,13 @@ GenTreePtr Compiler::fgMorphCall(GenTreeCall* call)
 #ifdef DEBUG
         // Tail call needs to be in one of the following IR forms
         //    Either a call stmt or
-        //    GT_RETURN(GT_CALL(..)) or
-        //    var = call
-        noway_assert((stmtExpr->gtOper == GT_CALL && stmtExpr == call) ||
-                     (stmtExpr->gtOper == GT_RETURN &&
-                      (stmtExpr->gtOp.gtOp1 == call || stmtExpr->gtOp.gtOp1->gtOp.gtOp1 == call)) ||
-                     (stmtExpr->gtOper == GT_ASG && stmtExpr->gtOp.gtOp2 == call));
+        //    GT_RETURN(GT_CALL(..)) or GT_RETURN(GT_CAST(GT_CALL))
+        //    var = GT_CALL or var = (GT_CAST(GT_CALL)
+        noway_assert(
+            (stmtExpr->gtOper == GT_CALL && stmtExpr == call) ||
+            (stmtExpr->gtOper == GT_RETURN &&
+             (stmtExpr->gtOp.gtOp1 == call || stmtExpr->gtOp.gtOp1->gtOp.gtOp1 == call)) ||
+            (stmtExpr->gtOper == GT_ASG && (stmtExpr->gtOp.gtOp2 == call || stmtExpr->gtOp.gtOp2->gtOp.gtOp1 == call)));
```

Comment:
In both of the above places we also need to assert that it is not only GT_CAST but a non-overflow checking one.  Otherwise, it is not correct to remove GT_CAST.